### ISSUE TITLE
Multi GPU for large model

### ISSUE THREAD

**Issue by neofob** [imartinez/privateGPT#1390]
Does privateGPT support multi-gpu for loading model that does not fit into one GPU?

For example, the Mistral 7B model requires 24 GB VRAM. Would having 2 Nvidia 4060 Ti 16GB help?

Thanks!

**Comment by charlyjna** [comment]
I tried but the application start with 2 graphic cards. I have to set a variable to force using only one graphic card (I have 2 x Quadro RTX 6000):

`CUDA_VISIBLE_DEVICES=0 poetry run python -m private_gpt`

**Comment by neofob** [comment]
@charlyjna : Multi-GPU crashes on "Query Docs" mode for me as well.
It works in "LLM Chat" mode though.
I have a RTX 4000 Ada SSF and a P40.

**Comment by jannikmi** [comment]
Is there any update on this?
I see the following error in the Frontend in "Query Docs" mode when using multipe GPUs: `error Expected all tensors to be on the same device, but found at least two devices, cuda:3 and cuda:0! (when checking argument for argument index in method wrapper_CUDA__index_select)
`

As @neofob mentioned, "LLM Chat" mode works.

**Comment by Luis-Sarabando** [comment]
Looking for an update as well on this, 

```
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:7 and cuda:0! (when checking argument for argument index in method wrapper_CUDA__index_select)
Generating embeddings:   0%|                                                                                                                                                                                                    | 0/1 [00:18<?, ?it/s]
make: *** [Makefile:52: ingest] Error 1
```

**Comment by mrharrison007** [comment]
> I tried but the application crashes with 2 graphic cards. I have to set a variable to force using only one graphic card (I have 2 x Quadro RTX 6000):
> 
> `CUDA_VISIBLE_DEVICES=0 poetry run python -m private_gpt`

Thank you for the "CUDA_VISIBLE_DEVICES=0" intel, privateGPT did not know what to do with my other 99 GPUs.  Just kidding, I only have 2 total for now.  98 more on the pipe-dream list :)

**Comment by SuperSonnix71** [comment]
https://github.com/imartinez/privateGPT/issues/1521#issuecomment-1963085662

**Comment by neofob** [comment]
> > @charlyjna : Multi-GPU crashes on "Query Docs" mode for me as well. It works in "LLM Chat" mode though. I have a RTX 4000 Ada SSF and a P40.
> 
> do you need to modify any settings.yaml to use Multi-GPU?

Nope, no need to modify settings.yaml. Multi-GPU works right out of the box in chat mode atm.

**Comment by thanhtantran** [comment]
> > > @charlyjna : Multi-GPU crashes on "Query Docs" mode for me as well. It works in "LLM Chat" mode though. I have a RTX 4000 Ada SSF and a P40.
> > 
> > 
> > do you need to modify any settings.yaml to use Multi-GPU?
> 
> Nope, no need to modify settings.yaml. Multi-GPU works right out of the box in chat mode atm.

Multi-GPU increases buffer size to GPU or not? 

I got the buffer size to GPU same for 1GPU and 2GPU

**Comment by davedcne** [comment]
I am also seeing this problem with tensors expected on one device for document ingest / query. Any update on this issue?

**Comment by neofob** [comment]
@davedcne : We have a related issue #1521 (Ingestion in parallel with multi-GPU) with pending patch for it. Please try it out and let us know.

**Comment by neofob** [comment]
**Recap:**
* [x]: Multi-GPU in LLM Chat mode works
* [x]: Need to apply patch from issue #1521 ; [patch][0] from @SuperSonnix71

[0]: https://gist.github.com/SuperSonnix71/e6ed139497f5176a68960db529e06163
